subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c

```diff
@@ -0,0 +1,1303 @@
+		if ((len < 0) || (len > sizeof(topic))) {
+			LOG_ERR("Failed to create twin report topic");
+			return -ENOMEM;
+		}
+		break;
+	case AZURE_IOT_HUB_TOPIC_TWIN_REQUEST:
+		len = snprintk(topic, sizeof(topic),
+			       TOPIC_TWIN_REQUEST, k_uptime_get_32());
+		if ((len < 0) || (len > sizeof(topic))) {
+			LOG_ERR("Failed to create device twin request");
+			return -ENOMEM;
+		}
+		break;
+	default:
+		LOG_ERR("No topic specified");
+		return -ENOMSG;
+	}
+
+	param.message.topic.topic.size = len;
+	param.message.topic.topic.utf8 = topic;
+
+	LOG_DBG("Publishing to topic: %s",
+		log_strdup(param.message.topic.topic.utf8));
+
+	return mqtt_publish(&client, &param);
+}
+
+int azure_iot_hub_disconnect(void)
+{
+	int err;
+
+	if (connection_state != STATE_CONNECTED) {
+		LOG_ERR("Azure IoT Hub is not connected");
+		return -EACCES;
+	}
+
+	connection_state = STATE_DISCONNECTING;
+
+	err = mqtt_disconnect(&client);
+	if (err) {
+		LOG_ERR("Failed to disconnect MQTT client, error: %d", err);
+		return err;
+	}
+
+	connection_state = STATE_INIT;
+
+	k_sleep(K_MSEC(100));
+
+	return 0;
+}
+
+int azure_iot_hub_connect(void)
+{
+	if (connection_state != STATE_INIT) {
+		LOG_ERR("Azure IoT Hub is not initialized");
+		return -EACCES;
+	}
+
+	return connect_client(&conn_config);
+}
+
+int azure_iot_hub_init(const struct azure_iot_hub_config *const config,
+		       const azure_iot_hub_evt_handler_t event_handler)
+{
+	int err;
+
+	if (connection_state != STATE_IDLE) {
+		LOG_ERR("Azure IoT Hub is already initialized");
+		return -EALREADY;
+	}
+
+	if (event_handler == NULL) {
+		LOG_ERR("Event handler must be provided");
+		return -EINVAL;
+	}
+
+	if (IS_ENABLED(CONFIG_AZURE_IOT_HUB_DEVICE_ID_APP)) {
+		if (config == NULL) {
+			LOG_ERR("Configuration must be provided");
+			return -EINVAL;
+		} else if (config->device_id == NULL) {
+			LOG_ERR("Client ID not set in the application");
+			return -EINVAL;
+		} else if (config->device_id_len >
+			   CONFIG_AZURE_IOT_HUB_DEVICE_ID_MAX_LEN) {
+			LOG_ERR("Device ID is too long, maximum length is %d",
+				CONFIG_AZURE_IOT_HUB_DEVICE_ID_MAX_LEN);
+			return -ENOMEM;
+		}
+
+		memcpy(conn_config.device_id, config->device_id,
+		       config->device_id_len);
+		conn_config.device_id_len = config->device_id_len;
+	}
+
+#if IS_ENABLED(CONFIG_AZURE_IOT_HUB_DPS)
+	struct dps_config cfg = {
+		.mqtt_client = &client,
+		.handler = dps_handler,
+		.reg_id = conn_config.device_id,
+		.reg_id_len = conn_config.device_id_len,
+	};
+
+	err = dps_init(&cfg);
+	if (err) {
+		LOG_ERR("Failed to initialize DPS, error: %d", err);
+		return err;
+	}
+
+	LOG_DBG("DPS initialized");
+#endif /* IS_ENABLED(CONFIG_AZURE_IOT_HUB_DPS) */
+
+#if IS_ENABLED(CONFIG_AZURE_FOTA)
+	err = azure_fota_init(fota_evt_handler);
+	if (err) {
+		LOG_ERR("Failed to initialize Azure FOTA, error: %d", err);
+		return err;
+	}
+
+	LOG_DBG("Azure FOTA initialized");
+#endif
+
+	evt_handler = event_handler;
+	is_initialized = true;
+	(void)err;
```

Comment:
why is err void casted here?

Reply:
That's because if DPS and FOTA are disabled, it will otherwise give an "unused variable" compiler warning